### PR TITLE
Bug 1798456: Make operator bundle for OML

### PR DESCRIPTION
**Pull request by ingvagabund**
TBD

TODO:
- [x] create valid CSV file
- [ ] honor manifest layout
- [ ] provider valid image-references file
- [x] make sure an operator bundle with role binding to external `extension-apiserver-authentication-reader` role (https://kubernetes.io/docs/tasks/access-kubernetes-api/#extension-apiserver-authenticates-the-request) is properly created
- [x] provider package.yaml file
- [x] The Dockerfile must create the /manifests directory at the top level of the image.
- [x] deploy the operator through OLM to verify all bits are properly set and available

Links:
- how to build CSV file: https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md
- CSV file example: https://github.com/openshift/sriov-network-operator/blob/master/manifests/4.4/sriov-network-operator.v4.4.0.clusterserviceversion.yaml
- manifest directory layout: https://github.com/operator-framework/operator-registry/tree/release-4.2#manifest-format

/hold

### CODE REVIEW

**Comment by ingvagabund** [comment]
In order to test OLM is able to create the deployment as specified in the CSV I had to:
- create SA, ClusterRole and ClusterRoleBinding as is specified in the CSV
- create operatorgroup:
  ```yaml
  apiVersion: operators.coreos.com/v1
  kind: OperatorGroup
  metadata:
    name: openshift-cluster-descheduler-operator
    namespace: openshift-descheduler-operator
  spec:
    targetNamespaces:
      - openshift-descheduler-operator
  ```
After that:
```sh
$ oc apply -f cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml
$ oc get deployment -n openshift-descheduler-operator
NAME                   READY     UP-TO-DATE   AVAILABLE   AGE
descheduler-operator   1/1       1            1           104s
```

**Comment by ingvagabund** [comment]
Alright, I was able to register the operator through operator-registry and have it installed through the console. Though, I still had to create the same resources as in the previous comment. To summarize the steps:

1. update the following line in Dockerfile:
   ```
   COPY --from=builder /go/src/github.com/openshift/cluster-kube-descheduler-operator/manifests /manifests
   ```
   to
   ```
   COPY --from=builder /go/src/github.com/openshift/cluster-kube-descheduler-operator/manifests/4.4 /manifests
   ```
   I assume this step is done automatically in the ART pipeline (e.g. pointing to the version that is currently built)
2. build and push the bundle:
   ```sh
   $ podman build -t quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle:latest -f Dockerfile .
   $ podman push quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle:latest
   ```
3. build and push image index for operator-registry (don't forget to increase the number of open files, .e.g. `ulimit -n 100000`)
   ```sh
   $ ./bin/linux-amd64-opm index add --bundles quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle:latest --tag quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle-index:1.0.0
   $ podman push quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle-index:1.0.0
   ```
4. create and apply catalogsource manifest:
   ```yaml
   apiVersion: operators.coreos.com/v1alpha1
   kind: CatalogSource
   metadata:
     name: cluster-kube-descheduler-operator
     namespace: openshift-marketplace
   spec:
     sourceType: grpc
     image: quay.io/jchaloup/ose-cluster-kube-descheduler-operator-bundle-index:1.0.0
   ```
5. open the console

**Comment by ingvagabund** [comment]
After removing an `image-reference` file from the bundle I was able to deploy the operator \O/.

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      The Descheduler operator exposes the following parameters in its CRD:
+
+      * `deschedulingIntervalSeconds` - this sets the number of seconds between descheduler runs
+      * `image` - specifies the Descheduler container image to deploy
+      * `flags` - this allows additional descheduler flags to be set, and they will be appended to the descheduler pod. Therefore, they must be in the same format as would be passed to the descheduler binary (eg, `"--dry-run"`)
+  displayName: Descheduler Operator
+  keywords: ["scheduling", "descheduler", "workload"]
+  provider:
+    name: Red Hat, Inc.
+  maturity: alpha # or beta/stable maybe?
```
Comment (soltysh) [inline manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: beta

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+          },
+          "spec": {
+            "image": "gcr.io/k8s-staging-descheduler/descheduler:0.9.0-79-gd845040d7",
+            "deschedulingIntervalSeconds": 3600
+          }
+        }
+      ]
+    certified: "false"
+    containerImage: quay.io/openshift/origin-cluster-kube-descheduler-operator:4.4
+    createdAt: 2020/02/17 # TODO(jchaloup) update the date before the PR gets merged
```
Comment (soltysh) [inline manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: do you want to update that?

File: manifests/01_namespace.yaml [hunk 3]
```diff
@@ -1,4 +0,0 @@
-apiVersion: v1
```
Comment (soltysh) [inline manifests/01_namespace.yaml]: Why these are removed?

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+          verbs:
+          - "*"
+        - apiGroups:
+          - apps
+          resources:
+          - deployments
+          - replicasets
+          verbs:
+          - "*"
+        - apiGroups: ["batch", "extensions"]
```
Comment (soltysh) [inline manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: What we use extensions for? We should not.

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            matchLabels:
+              name: descheduler-operator
+          template:
+            metadata:
+              labels:
+                name: descheduler-operator
+            spec:
+              containers:
+                - name: descheduler-operator
+                  image: docker.io/mdame/cluster-kube-descheduler-operator
```
Comment (soltysh) [inline manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: That needs changing, right?

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml (hunk 4 above)
Comment (ingvagabund) [reply manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: We don't use cron jobs anymore after merging https://github.com/openshift/cluster-kube-descheduler-operator/pull/66/commits/b64e7c0cdb499fb52e02ed51e92f2e9d55e1ccb5?file-filters%5B%5D=.go#diff-ad0243c09cf8e6c22f5f54bd27636f36L235 so we might remove entire rule.

File: manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml (hunk 5 above)
Comment (ingvagabund) [reply manifests/4.4/cluster-kube-descheduler-operator.v4.4.0.clusterserviceversion.yaml]: Right, just found https://quay.io/repository/openshift/origin-descheduler?tab=tags. Will update csv and retest.

File: manifests/01_namespace.yaml (hunk 3 above)
Comment (ingvagabund) [reply manifests/01_namespace.yaml]: They do not respect the manifest dir layout: https://github.com/operator-framework/operator-registry/tree/release-4.2#manifest-format

**Pull request merged**